### PR TITLE
Add objective-c codegen

### PR DESCRIPTION
**Pull request by umeshp7**
Objective-C codegen using https://developer.apple.com/documentation/foundation/nsurlsession.

### CODE REVIEW

File: codegens/objective-c/lib/index.js [hunk 1]
```diff
@@ -0,0 +1,285 @@
+var _ = require('./lodash'),
+  sanitizeOptions = require('./util').sanitizeOptions,
+  sanitize = require('./util').sanitize,
+  self;
+
+/**
+ * Parses Raw data to fetch syntax
```
Comment (shreys7) [inline codegens/objective-c/lib/index.js]: fetch syntax? 🤦‍♂

File: codegens/objective-c/lib/index.js [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+ *
+ * @param {Object} body Raw body data
+ * @param {String} indent Indent
```
Comment (shreys7) [inline codegens/objective-c/lib/index.js]: only Indent would not explain.

File: codegens/objective-c/lib/index.js [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+  'dataUsingEncoding:NSUTF8StringEncoding]];\n';
+  bodySnippet += '[request setHTTPBody:postData];\n';
+  return bodySnippet;
+}
+
+/**
+ * Parses GraphQL body
+ *
+ * @param {Object} body GraphQL body
+ * @param {String} indent Indent
```
Comment (shreys7) [inline codegens/objective-c/lib/index.js]: same as above

File: codegens/objective-c/lib/index.js [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+ * @param {Boolean} trim Trim request body if this is true.
+ */
+function parseGraphqlBody (body, indent, trim) {
```
Comment (shreys7) [inline codegens/objective-c/lib/index.js]: can we change this to parseGraphQLBody? or just parseGraphQL? We have used parseGraphQL everywhere else.

File: codegens/objective-c/lib/index.js [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+  var bodySnippet = '',
+    rawBody = JSON.stringify(body);
+  bodySnippet += 'NSData *postData = [[NSData alloc] initWithData:[@"' + sanitize(rawBody, trim) + '" ' +
+  'dataUsingEncoding:NSUTF8StringEncoding]];\n';
+  bodySnippet += '[request setHTTPBody:postData];\n';
+  return bodySnippet;
+}
+
+/**
+ * Parses URLEncoded body from request to fetch syntax
```
Comment (shreys7) [inline codegens/objective-c/lib/index.js]: fetch syntax

File: codegens/objective-c/lib/index.js [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+      }
+      first = false;
+    }
+  });
+  bodySnippet += '[request setHTTPBody:postData];\n';
+  return bodySnippet;
+}
+
+/**
+ * Parses URLEncoded body from request to fetch syntax
```
Comment (shreys7) [inline codegens/objective-c/lib/index.js]: fetch syntax

File: codegens/objective-c/lib/index.js [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+ *
+ * @param {Object} body URLEncoded Body
+ * @param {String} indent Indent
+ * @param {Boolean} trim Trim request body is this true.
```
Comment (shreys7) [inline codegens/objective-c/lib/index.js]: Trim request body is this true?

File: codegens/objective-c/lib/index.js [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+    first = true;
+
+  _.forEach(body, function (data) {
+    key = trim ? data.key.trim() : data.key;
+    value = trim ? data.value.trim() : data.value;
+    if (!data.disabled) {
+      if (first) {
+        bodySnippet += 'NSArray *parameters = @[';
+      }
+      if (!first) {
```
Comment (shreys7) [inline codegens/objective-c/lib/index.js]: Can we use `array.join(',')` syntax? Store all the individual params in an array and then at last use .join()?

File: codegens/objective-c/lib/index.js [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+  bodySnippet += 'NSData *postData = [body dataUsingEncoding:NSUTF8StringEncoding];\n';
+  bodySnippet += '[request setHTTPBody:postData];\n';
+  return bodySnippet;
+}
+
+/**
+ * Parses Body from the Request
+ *
+ * @param {Object} body body object from request.
+ * @param {String} indent
```
Comment (shreys7) [inline codegens/objective-c/lib/index.js]: jsdocs

File: codegens/objective-c/lib/index.js [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+  if (!_.isEmpty(body)) {
+    switch (body.mode) {
+      case 'urlencoded':
+        return parseURLEncodedBody(body.urlencoded, indent, trim);
+      case 'raw':
+        return parseRawBody(body.raw, indent, trim);
+      case 'formdata':
+        return parseFormData(body.formdata, indent, trim);
+      case 'file':
+        return '';
```
Comment (shreys7) [inline codegens/objective-c/lib/index.js]: you could return a `'<file-content-here>'` as we do in all other codegens.

File: codegens/objective-c/lib/index.js [hunk 11]
```diff
@@ -0,0 +1,285 @@
...
+    }
+  }
+  return '';
+}
+
+/**
+ * Parses headers from the request.
+ *
+ * @param {Object} obj
```
Comment (shreys7) [inline codegens/objective-c/lib/index.js]: jsdocs

File: codegens/objective-c/lib/index.js [hunk 12]
```diff
@@ -0,0 +1,285 @@
...
+ */
+function parseHeaders (obj, indent, trim) {
+  let headers = '';
+  if (_.isEmpty(obj)) {
+    return headers;
+  }
+  headers = indent + 'NSDictionary *headers = @{\n';
+  var first = true;
+  _.forEach(obj, function (value, key) {
+    if (!first) {
```
Comment (shreys7) [inline codegens/objective-c/lib/index.js]: same comment. `.join()` syntax

File: codegens/objective-c/lib/index.js [hunk 13]
```diff
@@ -0,0 +1,285 @@
...
+    let obj = {};
+    codeSnippet = '#import <Foundation/Foundation.h>\n';
+    codeSnippet += 'NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:[NSURL URLWithString:@"' +
+      encodeURI(request.url.toString()) + '"]\n';
+    codeSnippet += `${indent}cachePolicy:NSURLRequestUseProtocolCachePolicy\n`;
+    codeSnippet += `${indent}timeoutInterval:${requestTimeout}.0];\n`;
+
+    // TODO: use defaultSessionConfiguration
+    // codeSnippet += 'NSURLSessionConfiguration *config = [NSURLSessionConfiguration defaultSessionConfiguration];\n';
+
```
Comment (shreys7) [inline codegens/objective-c/lib/index.js]: Add multiple/no/single file handling code here

File: codegens/objective-c/lib/index.js [hunk 14]
```diff
@@ -0,0 +1,285 @@
...
+  }
+  headers = indent + 'NSDictionary *headers = @{\n';
+  var first = true;
+  _.forEach(obj, function (value, key) {
+    if (!first) {
+      headers += ',\n';
+    }
+    first = false;
+    if (typeof value !== 'string') {
+      headers += indent.repeat(2) + '@"' + key + '":@"' + value + '"';
```
Comment (shreys7) [inline codegens/objective-c/lib/index.js]: Will this work for value an array?

File: codegens/objective-c/test/unit/convert.test.js [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+        expect.fail(null, null, error);
+      }
+      expect(snippet).to.be.a('string');
+      var snippetArray = snippet.split('\n'),
+        i;
+      for (i = 0; i < snippetArray.length; i++) {
+        if (snippetArray[i].startsWith('cachePolicy:NSURLRequestUseProtocolCachePolicy')) {
+          expect(snippetArray[i + 1].charAt(0)).to.equal(' ');
+          expect(snippetArray[i + 1].charAt(1)).to.equal(' ');
+          expect(snippet).to.include('timeoutInterval:10');
```
Comment (shreys7) [inline codegens/objective-c/test/unit/convert.test.js]: Is 10 used as infinity?

File: codegens/objective-c/lib/index.js (hunk 13 above)
Comment (umeshp7) [reply codegens/objective-c/lib/index.js]: Meaning?

File: codegens/objective-c/test/unit/convert.test.js (hunk 15 above)
Comment (umeshp7) [reply codegens/objective-c/test/unit/convert.test.js]: I couldn't find a way to set infinite request time out for nsurlsession. So setting a 10 sec convention I found in many sample snippets used.

File: codegens/objective-c/lib/index.js (hunk 13 above)
Comment (umeshp7) [reply codegens/objective-c/lib/index.js]: This TODO is different from that. 
I have added the file handling elsewhere.

File: codegens/objective-c/lib/index.js (hunk 4 above)
Comment (umeshp7) [reply codegens/objective-c/lib/index.js]: Done

File: codegens/objective-c/lib/index.js (hunk 13 above)
Comment (shreys7) [reply codegens/objective-c/lib/index.js]: Actually what I meant was add the file handling code we added before parsrBody fiunction call.

File: codegens/objective-c/test/unit/convert.test.js (hunk 15 above)
Comment (shreys7) [reply codegens/objective-c/test/unit/convert.test.js]: Ah, okay!

File: codegens/objective-c/lib/index.js (hunk 10 above)
Comment (shreys7) [reply codegens/objective-c/lib/index.js]: Not here @umeshp7 . I meant to add this placeholder for case 'file'

File: README.md [hunk 16]
```diff
@@ -29,6 +29,7 @@ List of supported code generators:
 | NodeJs | Native |
 | NodeJs | Request |
 | NodeJs | Unirest |
+| Objective-C| NSURLSession|
```
Comment (abhijitkane) [inline README.md]: space b/w `-C` and `|`?

Comment (umeshp7) [reply README.md]: In the preview it doesn't make a difference.

**Pull request merged**